torch/distributed/tensor/experimental/_attention.py

```diff
@@ -62,6 +72,171 @@ class _ContextParallelOptions:
+    """
+
+    ROUND_ROBIN_CYCLE = 2
+
+    @classmethod
+    def shard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+    ) -> torch.Tensor:
+        assert cls.ROUND_ROBIN_CYCLE == 2, (
+            "The current implementation only works if ROUND_ROBIN_CYCLE is 2 "
+            f"but got {cls.ROUND_ROBIN_CYCLE}"
+        )
+        cp_world_size = mesh.size()
+        cp_rank = mesh.get_local_rank()
+        assert buffer.size()[seq_dim] % (cp_world_size * 2) == 0
+        chunks = buffer.chunk(cp_world_size * 2, dim=seq_dim)
+        return torch.cat(
+            (chunks[cp_rank], chunks[cp_world_size * 2 - cp_rank - 1]),
+            dim=seq_dim,
+        )
+
+    @classmethod
+    def unshard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+    ) -> torch.Tensor:
+        assert cls.ROUND_ROBIN_CYCLE == 2, (
+            "The current implementation only works if ROUND_ROBIN_CYCLE is 2 "
+            f"but got {cls.ROUND_ROBIN_CYCLE}"
+        )
+        buffer = buffer.contiguous()
+        cp_world_size = mesh.size()
+
+        all_buffers = [torch.empty_like(buffer) for _ in range(cp_world_size)]
+        ft_c.all_gather_inplace(all_buffers, buffer, mesh)
+        sliced_buffers = [sb for b in all_buffers for sb in b.chunk(2, dim=seq_dim)]
+        ordered_buffers = list(sliced_buffers)
+        for i, b in enumerate(sliced_buffers):
+            if i % 2 == 0:
+                ordered_buffers[i // 2] = b
+            else:
+                ordered_buffers[cp_world_size * 2 - (i // 2) - 1] = b
+        return torch.cat(ordered_buffers, dim=seq_dim)
+
+
+# TODO: merge with class `_LoadBalancer`
+class FlexAttentionSharder(ABC):
+    @abstractmethod
+    def shard(
+        self,
+        tensor: torch.Tensor,
+        num_chunks: int,
+        shard_dim: int,
+        **kwargs: dict[str, Any],
+    ) -> list[torch.Tensor]: ...
+
+    @abstractmethod
+    def unshard(
+        self, shards: list[torch.Tensor], shard_dim: int, **kwargs: dict[str, Any]
+    ) -> torch.Tensor: ...
+
+    @abstractmethod
+    @lru_cache  # noqa: B019
+    def gen_block_sharding_map(
+        self, num_chunks: int, block_mask: BlockMask, **kwargs: dict[str, Any]
+    ) -> torch.Tensor: ...
+
+
+class FlexAttentionContiguousSharder(FlexAttentionSharder):
+    def shard(
+        self,
+        tensor: torch.Tensor,
+        num_chunks: int,
+        shard_dim: int,
+        **kwargs: dict[str, Any],
+    ) -> list[torch.Tensor]:
+        return list(tensor.chunk(num_chunks, dim=shard_dim))
+
+    def unshard(
+        self, shards: list[torch.Tensor], shard_dim: int, **kwargs: dict[str, Any]
+    ) -> torch.Tensor:
+        return torch.cat(shards, dim=shard_dim)
+
+    @lru_cache  # noqa: B019
+    def gen_block_sharding_map(
+        self, num_chunks: int, block_mask: BlockMask, **kwargs: dict[str, Any]
+    ) -> torch.Tensor:
+        block_mask_tuple = block_mask.as_tuple()
+        assert len(block_mask_tuple) == 13
```

Comment:
nit: is there a good reason to operate on the tuple as opposed to the named members?

Reply:
Good catch. I think I only need `as_tuple()` when need to pass the block_mask to HOP call, and can use named members in other places.